ext/sockets/sockets.c

```diff
@@ -2755,6 +2774,117 @@ PHP_FUNCTION(socket_addrinfo_explain)
 }
 /* }}} */
 
+#ifdef PHP_WIN32
+
+#include "ext/standard/base64.h"
+
+/* {{{ proto string socket_wsaprotocol_info_export(resource stream, int target_pid)
+   Exports the network socket information suitable to be used in another process. */
+PHP_FUNCTION(socket_wsaprotocol_info_export)
+{
+	WSAPROTOCOL_INFO wi;
+	zval *zsocket;
+	php_socket *socket;
+	zend_long target_pid;
+	zend_string *ret;
+	zend_off_t offset = 0;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "rl", &zsocket, &target_pid) == FAILURE) {
+		return;
+	}
+	if ((socket = (php_socket *) zend_fetch_resource(Z_RES_P(zsocket), le_socket_name, le_socket)) == NULL) {
+		RETURN_FALSE;
+	}
+
+	if (SOCKET_ERROR == WSADuplicateSocket(socket->bsd_socket, (DWORD)target_pid, &wi)) {
+		DWORD err = WSAGetLastError();
+		LPSTR buf = NULL;
+
+		if (!FormatMessage(
+				FORMAT_MESSAGE_ALLOCATE_BUFFER |
+				FORMAT_MESSAGE_FROM_SYSTEM |
+				FORMAT_MESSAGE_IGNORE_INSERTS,
+				NULL,
+				err,
+				MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+				(LPSTR)&buf,
+			0, NULL)) {
+			php_error_docref(NULL, E_WARNING, "Unable to export WSA protocol info [0x%08lx]", err);
+		} else {
+			php_error_docref(NULL, E_WARNING, "Unable to export WSA protocol info [0x%08lx]: %s", err, buf);
+		}
+
+		RETURN_NULL();
+	}
+
+	/* The simplest way to serialize. The internal data might be of interest,
+		however it is unlikely to ever make sense to have it editable within
+		a script. The struct is suited very well and has same size under both
+		64- and 32-bit. */
+	ret = php_base64_encode((const char *)&wi, sizeof(wi));
+
+	RETVAL_STR((ret));
+
+	zend_string_free(ret);
```

Comment:
Isn't this freeing the string that was just returned?

Reply:
Ohh, my bad. It's just an assignment there. Fixing.